src/render/program/symbol_program.js

```diff
@@ -196,12 +214,29 @@ const symbolIconUniformValues = (
         'u_is_text': +isText,
         'u_pitch_with_map': +pitchWithMap,
         'u_texsize': texSize,
-        'u_tile_id': tileID,
-        'u_zoom_transition': zoomTransition,
-        'u_inv_rot_matrix': invRotMatrix,
-        'u_merc_center': mercCenter,
-        'u_texture': 0
+        'u_texture': 0,
+        'u_tile_id': [0, 0, 0],
+        'u_zoom_transition': 0,
+        'u_inv_rot_matrix': identityMatrix,
+        'u_merc_center': [0, 0],
+        'u_camera_forward': [0, 0, 0],
+        'u_ecef_origin': [0, 0, 0],
+        'u_tile_matrix': identityMatrix
     };
+
+    if (transform.projection.name === 'globe') {
+        const tileMatrix = calculateGlobeMatrix(transform, transform.worldSize);
+        const id = coord.toUnwrapped();
+        values['u_tile_id'] = [coord.canonical.x, coord.canonical.y, 1 << coord.canonical.z];
+        values['u_zoom_transition'] = zoomTransition;
+        values['u_inv_rot_matrix'] = tileTransform.createInversionMatrix(id);
+        values['u_merc_center'] = mercatorCenter;
+        values['u_camera_forward'] = [cameraForward[0], cameraForward[1], cameraForward[2]];
```

Comment:
@mourner I'd prefer to avoid creating an extraneous array here, but this was the way I found to make flow work post https://github.com/mapbox/mapbox-gl-js/pull/11400 (Refer to commit https://github.com/mapbox/mapbox-gl-js/pull/11399/commits/c362729c3770725d748674275f7451271926ca49 about the change that had to be made). Do you have any tip on how to simply have `transform._camera.forward()` as it was set before (which returns a `Vec3`? I tried a few different ways but the flow messages weren't really helpful.

Reply:
@karimnaaji before the change, `forward()` was essentially typed as `any`, so Flow always passed silently regardless of how it was used. Now it's typed, but because of incompatibilities between tuple types and arrays in Flow and the way uniforms are currently typed, Flow can't allow passing Vec3 as a uniform and there's no easy way around it. For now, we can [explicitly convert it through any](https://github.com/mapbox/mapbox-gl-js/pull/11399/commits/b8c75ecd30d60a1d2be9d76e78aa41bfd10a9343) to avoid temp allocation, but we'll try to improve typing around this in the future — hopefully it'll be easier to do in more recent versions of Flow (the current ancient one has a lot of problems).